packages/firestore/src/api/user_data_reader.ts

```diff
@@ -504,13 +506,16 @@ export function parseUpdateVarargs(
   for (let i = keys.length - 1; i >= 0; --i) {
     if (!fieldMaskContains(fieldMaskPaths, keys[i])) {
       const path = keys[i];
-      const value = values[i];
+      let value = values[i];
+
+      // For Compat types, we have to "extract" the underlying types before
+      // performing validation.
+      if (value instanceof Compat) {
```

Comment:
I'd like to see us get to a place where we're not wrapping and unwrapping these types but instead just sharing a single definition for all the variations. Obviously this isn't a topic for this PR and I'm not advocating changes right now.

Reply:
I don't quite see how this would work, but would love to get there. Maybe we can talk about this more, since I would need some guidance on how to achieve this goal.

Reply:
Let's chat offline; no need to address in this PR anyway.